src/main/java/sparksoniq/spark/closures/ForClauseFlatMapClosure.java

```diff
@@ -20,36 +20,35 @@
 package sparksoniq.spark.closures;
 
 import org.apache.spark.api.java.function.FlatMapFunction;
-import org.apache.spark.api.java.function.Function;
-import org.apache.spark.api.java.function.MapFunction;
 import org.apache.spark.sql.Row;
-import org.apache.spark.sql.RowFactory;
 import org.apache.spark.sql.types.StructType;
 import sparksoniq.jsoniq.item.Item;
 import sparksoniq.jsoniq.runtime.iterator.RuntimeIterator;
 import sparksoniq.semantics.DynamicContext;
-import sparksoniq.spark.SparkSessionManager;
 
 import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
 
 public class ForClauseFlatMapClosure implements FlatMapFunction<Row, Row> {
-    RuntimeIterator _expression;
+    private RuntimeIterator _expression;
+    private StructType _oldSchema;
+    private int _duplicateColumnIndex;
 
-    public ForClauseFlatMapClosure(RuntimeIterator expression) {
+    public ForClauseFlatMapClosure(RuntimeIterator expression, StructType oldSchema, int duplicateVariableIndex) {
         this._expression = expression;
+        this._oldSchema = oldSchema;
+        this._duplicateColumnIndex = duplicateVariableIndex;
     }
 
     @Override
     public Iterator<Row> call(Row row) {
-        StructType schema = row.schema();
-        String[] columnNames = schema.fieldNames();
+        String[] columnNames = _oldSchema.fieldNames();
 
         // Deserialize row
         List<Object> deserializedRow = ClosureUtils.deserializeEntireRow(row);
         List<List<Item>> rowColumns = new ArrayList<>();
```

Comment:
I have moved actually decided to make these changes out of curiosity. After moving all these new statements into private variables of the closure, the performance is nearly identical to the non-dataframe implementation. Results can be seen in the performance repo at https://github.com/CanBerker/sparksoniq-dataframe-perf-v1  (the jars are also available)

Reply:
I realized that I had forgotten the custom partition number of json-file in the initial run. I have also added this to the query and got the medium - 5m test set working. Results are fairly close between DF and alpha (Dataframes are even faster than the alpha by about 7 seconds).

The full dataset didn't work though. I will increase partitioning and try again.

Reply:
This is very good news!

In the long-term we may want to introduce a default partitioning number for local execution, because this also confused the students. On HDFS, the partitions are built automatically to match 128MB blocks, but in a local execution only one partition is made no matter how big the file is.